### DISCUSSION TITLE
PK names -- developer or user-determined?

### DISCUSSION THREAD

**Discussion by ecoon** [amanzi/amanzi#672] [Ideas]
PKs are currently named by the User -- this is provided by the parameter list's name.  For instance, in the following ATS input spec:

```
  <ParameterList name="cycle driver" type="ParameterList">
    ...
    <ParameterList name="PK tree" type="ParameterList">
      <ParameterList name="Richards" type="ParameterList">
        <Parameter name="PK type" type="string" value="richards flow" />
      </ParameterList>
    </ParameterList>
  </ParameterList>
```

The PK's name is set to "Richards", but this could be anything provided by the user.

This is really only used in one place -- the VerboseObject, which uses the name of the PK to set the prefix in the log file, e.g. this run results in lines like:

```
Coordinator        |   Cycle = 0,  Time [days] = 0,  dt [days] = 1.157407407407407e-05
Coordinator        |   ----------------------------------------------------------------------
Richards           |   ----------------------------------------------------------------
Richards           |   Advancing: t0 = 0 t1 = 1 h = 1
Richards           |   ----------------------------------------------------------------
```
where we once again see the name "Richards" appearing as the line prefix.

@lipnikov  suggests that he would like to see more rigorous, developer-named names here.  This discussion will help determine best practice for this, and future work will fix all PKs to follow this.  Note that nearly all PKs simply construct their VO in PK.hh, but a few (Alquimia_PK, maybe other chem ones) currently hard-code their name.

**Comment by ecoon** [comment]
To me, the name most important requirement is that the name must be unique per PK instance.  This is not as trivial as one might think.  For instance, we use "surface balance general" PKs, which are effectively simple ODEs.  The current Alquimia_PK convention is given by: `Alquimia_PK:DOMAIN_NAME` -- I have several runs where this would result in several PKs having the same name, e.g. `SurfBalGeneral:surface`.

Another thing to consider -- while I understand the desire to allow a developer to debug a user's output file, I think it is more important that a USER can debug the user's output file.  Especially if we move to more "generic" PKs, using a hard-coded prefix (e.g. Alquimia_PK) will only confuse the user.  Image "Conservative_Parabolic_PDE:surface".  No user is going to realize that this PK is their overland flow PK.

I recognize that @lipnikov  is concerned with multiple stages -- ATS does not use stages so I don't know how this makes things harder.  Neither of the current models seem to consider stage in setting the prefix, do they?

**Reply by lipnikov** [reply]
Many classes have a verbose object. It seem logical to me that if this object defines a name for output, this clearly helps the user to debug user's output by changing this name. Also any default behavior (e.g. taking name from list's name) could be overwritten by adding a properly formatted verbose object.

**Reply by lipnikov** [reply]
In Amanzi, the converter between specs generates a unique name for a PK which could be quite long, e.g. "injection of liquid CO2 via well 27:coupled flow reactive transport fracture". I do not want this name to be printed on each line. I may prefer a more concise name as default name,  like "CoupledFlowRT:fracture" which could be the same for each time stage.  
Converter will add this short name to verbose object. User may remove it for debugging purpose and use bla-bla-bla instead.
